File: 62. Unique Paths.go
package main

import "fmt"

var cache map[Cell]int

type Cell struct {
	x, y int
}

func didHitCache(row, col int) bool {
	_, didFind := cache[Cell{row, col}]
	return didFind
}

func dynamicProgramming(m, n int) int {
	if didHitCache(m, n) {
		return cache[Cell{m, n}]
	} else if m == 1 && n == 1 {
		return 1
	} else if m == 0 || n == 0 {
		return 0
	} else {
		left := dynamicProgramming(m-1, n)
		right := dynamicProgramming(m, n-1)
		cache[Cell{m, n}] = left + right
		return cache[Cell{m, n}]
	}
}

func uniquePaths(m int, n int) int {
	cache = map[Cell]int{}
	return dynamicProgramming(m, n)
}

func main() {
	ans := uniquePaths(35, 35)
	fmt.Println(ans)
}

File: 62. Unique Paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        cache = dict()
        if m == 1 and n == 1: return 1

        def dfs(r,c):
            if (r,c) in cache: return cache[(r,c)]
            elif r == (m-1) and c == (n-1): return 0
            elif r == (m-2) and c == (n-1): return 1
            elif r > m or c > n or r < 0 or c < 0: return 0
            elif r == (m-1) and c == (n-2): return 1
            elif (r,c) in cache: return cache[(r,c)]
            cache[(r,c)] = dfs(r+1,c) + dfs(r,c+1)
            return cache[(r,c)]

        return dfs(0,0)